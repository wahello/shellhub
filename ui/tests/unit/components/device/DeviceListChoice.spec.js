import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import DeviceList from '@/components/device/DeviceList';

describe('DeviceList', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  let wrapper;

  const numberDevices = 2;
  const devicesSelected = [];

  const pagination = {
    groupBy: [],
    groupDesc: [],
    itemsPerPage: 10,
    multiSort: false,
    mustSort: false,
    page: 1,
    sortBy: [],
    sortDesc: [],
  };

  const devices = [
    {
      uid: 'a582b47a42d',
      name: '39-5e-2a',
      identity: {
        mac: '00:00:00:00:00:00',
      },
      info: {
        id: 'linuxmint',
        pretty_name: 'Linux Mint 19.3',
        version: '',
      },
      public_key: '----- PUBLIC KEY -----',
      tenant_id: '00000000',
      last_seen: '2020-05-20T18:58:53.276Z',
      online: false,
      namespace: 'user',
      status: 'accepted',
    },
    {
      uid: 'a582b47a42e',
      name: '39-5e-2b',
      identity: {
        mac: '00:00:00:00:00:00',
      },
      info: {
        id: 'linuxmint',
        pretty_name: 'Linux Mint 19.3',
        version: '',
      },
      public_key: '----- PUBLIC KEY -----',
      tenant_id: '00000001',
      last_seen: '2020-05-20T19:58:53.276Z',
      online: true,
      namespace: 'user',
      status: 'accepted',
    },
  ];

  // const devicesOffline = JSON.parse(JSON.stringify(devices));
  // devicesOffline[1].online = false;

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      devices,
      numberDevices,
      devicesSelected,
    },
    getters: {
      'devices/list': (state) => state.devices,
      'devices/getNumberDevices': (state) => state.numberDevices,
      'devices/getDevicesSelected': (state) => state.devicesSelected,
    },
    actions: {
      'devices/setDevicesSelected': () => {},
      'devices/getDevicesMostUsed': () => {},
      'devices/fetch': () => {},
      'devices/resetListDevices': () => {},
      'snackbar/showSnackbarCopy': () => {},
      'snackbar/showSnackbarErrorAssociation': () => {},
      'snackbar/showSnackbarErrorLoading': () => {},
    },
  });

  ///////
  // In this case, it is tested when device is online.
  ///////

  describe('Suggested devices', () => {
    beforeEach(() => {
      wrapper = mount(DeviceList, {
        store,
        localVue,
        stubs: ['fragment', 'router-link'],
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.hostname).toEqual('localhost');
      expect(wrapper.vm.pagination).toEqual(pagination);
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.getListDevices).toEqual(devices);
      expect(wrapper.vm.getNumberDevices).toEqual(numberDevices);
    });
    it('Process data in methods', () => {
      Object.keys(devices).forEach((device) => {
        const address = `${device.namespace}.${device.name}@localhost`;
        expect(wrapper.vm.address(device)).toEqual(address);
      });
    });

    //////
    // HTML validation
    //////

    it('Renders the template with components', async () => {
      expect(wrapper.find('[data-test="deviceIcon-component"]').exists()).toEqual(true);
    });
    it('Renders the template with data', () => {
      const dt = wrapper.find('[data-test="dataTable-field"]');
      const dataTableProps = dt.vm.$options.propsData;

      expect(dataTableProps.items).toHaveLength(numberDevices);
    });
  });
});
