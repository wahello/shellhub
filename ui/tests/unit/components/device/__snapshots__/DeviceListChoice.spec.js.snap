// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceList Suggested devices Renders the component 1`] = `
<fragment-stub>
  <div class="v-card__text pa-0">
    <div class="v-data-table elevation-1 v-data-table--has-bottom theme--light" data-test="dataTable-field">
      <div class="v-data-table__wrapper">
        <table>
          <colgroup>
            <col class="">
            <col class="">
            <col class="">
            <col class="">
            <col class="">
            <col class="">
          </colgroup>
          <thead class="v-data-table-header">
            <tr>
              <th role="columnheader" scope="col" aria-label="Online: Not sorted. Activate to sort ascending." aria-sort="none" class="text-center sortable"><span>Online</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Hostname: Not sorted. Activate to sort ascending." aria-sort="none" class="text-center sortable"><span>Hostname</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Tags: Not sorted. Activate to sort ascending." aria-sort="none" class="text-center sortable"><span>Tags</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Operating System" class="text-center"><span>Operating System</span></th>
              <th role="columnheader" scope="col" aria-label="SSHID" class="text-center"><span>SSHID</span></th>
              <th role="columnheader" scope="col" aria-label="Actions" class="text-center"><span>Actions</span></th>
            </tr>
          </thead>
          <tbody>
            <tr class="">
              <td class="text-center"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">check_circle</i><span class="v-tooltip v-tooltip--bottom"><!----></span></td>
              <td class="text-center">
                <router-link-stub to="[object Object]">
                  39-5e-2a
                </router-link-stub>
              </td>
              <td class="text-center"></td>
              <td class="text-center">
                <fragment-stub data-test="deviceIcon-component"><i aria-hidden="true" data-test="type-icon" class="v-icon notranslate icons fl fl-linuxmint theme--light"></i></fragment-stub>
                Linux Mint 19.3
              </td>
              <td class="text-center"><span class="list-itens v-chip v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          user.39-5e-2a@localhost
          <button type="button" class="v-icon notranslate v-icon--link v-icon--right mdi mdi-content-copy theme--light" style="font-size: 16px;"></button></span></span></td>
              <td class="text-center"><button type="button" class="v-icon notranslate icons v-icon--link material-icons theme--light">info</button><span class="v-tooltip v-tooltip--bottom"><!----></span>
                <fragment-stub data-test="tagFormDialog-component"><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-tag theme--light"></button><span class="v-tooltip v-tooltip--bottom"><!----></span>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </fragment-stub>
                <!---->
                <fragment-stub data-test="deviceDelete-component"><span><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">delete</button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </fragment-stub>
              </td>
            </tr>
            <tr class="">
              <td class="text-center"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light success--text">check_circle</i></td>
              <td class="text-center">
                <router-link-stub to="[object Object]">
                  39-5e-2b
                </router-link-stub>
              </td>
              <td class="text-center"></td>
              <td class="text-center">
                <fragment-stub data-test="deviceIcon-component"><i aria-hidden="true" data-test="type-icon" class="v-icon notranslate icons fl fl-linuxmint theme--light"></i></fragment-stub>
                Linux Mint 19.3
              </td>
              <td class="text-center"><span class="list-itens v-chip v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          user.39-5e-2b@localhost
          <button type="button" class="v-icon notranslate v-icon--link v-icon--right mdi mdi-content-copy theme--light" style="font-size: 16px;"></button></span></span></td>
              <td class="text-center"><button type="button" class="v-icon notranslate icons v-icon--link material-icons theme--light">info</button><span class="v-tooltip v-tooltip--bottom"><!----></span>
                <fragment-stub data-test="tagFormDialog-component"><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-tag theme--light"></button><span class="v-tooltip v-tooltip--bottom"><!----></span>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </fragment-stub>
                <fragment-stub data-test="terminalDialog-component"><button type="button" class="v-icon notranslate ml-1 v-icon--link mdi mdi-console theme--light"></button><span class="v-tooltip v-tooltip--bottom"><!----></span>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </fragment-stub>
                <fragment-stub data-test="deviceDelete-component"><span><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">delete</button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </fragment-stub>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div class="v-data-footer">
        <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-121" class="v-input__slot">
                <div class="v-select__slot">
                  <div class="v-select__selections">
                    <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-121" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                  </div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                  </div><input type="hidden" value="10">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="v-data-footer__pagination">1-2 of 2</div>
        <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
        <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
      </div>
    </div>
  </div>
</fragment-stub>
`;
